scrolling

TheTrio opened the pull request:
# What

This PR aims to address #6. It adds a scrolling functionality to the text field, so that the user isn't overwhelmed by having all the words displayed at once. 

# Why

Besides being a better user experience, rendering fewer words at a single time makes the application faster since there aren't as many characters to render/keep track of. 

# How

The difficult part was figuring out when the user is at the end of a line. I don't think there's a way to do this using `rust-tui` so I've had to do the math myself. This logic is present in the `get_skip_count` function. Once we have calculated the amount of characters to skip, the implementation is relatively straightforward. 

# Checklist

- [ ] Documentation added
- [ ] Tests added
- [ ] No linting errors / broken tests
- [ ] Merge ready

Once this is merged, I think it would be nice to increase the `HORIZONTAL_MARGIN`  (dynamically) to center the text. It would result in a better UI in my opinion.

TheTrio committed:
implemented basic scrolling

TheTrio committed:
merge main

TheTrio committed:
make the current line always the 2nd

TheTrio commented:
Here's how it looks now

![demo](https://user-images.githubusercontent.com/10794178/166830707-a50c3a11-7c37-4908-942e-271556163993.gif)

coloradocolby commented:
Wow great progress! I'm out of town for a few days visiting family but will take a look at this when I get back home and settled in!

coloradocolby committed:
readme update

TheTrio commented:
hey @coloradocolby could you take a look at this?

coloradocolby commented:
Yes — I will take a look today! Thank you for the reminder 🙏

TheTrio committed:
Merge branch 'main' into develop

coloradocolby commented on src/main.rs:
```diff
@@ -31,7 +32,7 @@ const TICK_RATE_MS: u64 = 100;
 pub struct Cli {
     /// number of words to use in test
-    #[clap(short = 'w', long, default_value_t = 15)]
+    #[clap(short = 'w', long, default_value_t = 2000)]
```
Let's revert this to 15 before we merge. In fact if you want to add 
`const DEFAULT_NUM_WORDS: u64 = 15;` under the `TICK_RATE_MS` const above that would be great!

coloradocolby commented on src/thok.rs:
```diff
@@ -151,7 +158,7 @@ impl Thok {
 
     pub fn backspace(&mut self) {
-        if self.cursor_pos > 0 {
+        if self.cursor_pos > 0 && self.cursor_pos > self.total_line_length {
```
```suggestion
        if self.cursor_pos > self.total_line_length {
```
I believe you should be good to shorten this

coloradocolby commented on src/thok.rs:
```diff
@@ -234,4 +241,34 @@ impl Thok {
         Ok(())
     }
+
+    pub fn get_skip_count(&mut self, max_width: usize) {
```
Would you mind adding some comments in this fn?

TheTrio replied:
Added. Its been a while since I've looked at the code myself, but hopefully the comments explain the algorithm well enough.

coloradocolby commented on src/ui.rs:
```diff
@@ -54,22 +52,22 @@ impl Widget for &Thok {
                             ),
                             Constraint::Length(time_left_lines),
-                            Constraint::Length(prompt_occupied_lines),
+                            Constraint::Length(3),
```
should this be at most 3? If a prompt only takes up 1 line we shouldn't give it 3

TheTrio replied:
I don't think it actually matters since those lines wouldn't be visible would they? Its just empty space.

jrnxf reviewed (changes requested):
Left a few comments!

TheTrio committed:
remove redundant check

Co-authored-by: Colby Thomas <coloradocolby@gmail.com>

TheTrio committed:
Added comments. Reverted word count

coloradocolby commented:
> Also, there are some restrictions presently and I think they make sense but I'd like to hear your opinion as well -
> 
> 1. Once you've typed a line correctly and have moved onto the next line, you **CANNOT** move back to the previous line.
> 2. The program won't scroll until you have fixed all your mistakes in the frame(which presently spans 3 lines)
> 
> There's another issue which isn't really related but it might make sense to fix it before this gets merged - if you type a character other than a space, it isn't displayed in red(because there isn't anything to display). Earlier, it just meant that your accuracy would go down.
> 
> But with this, since you can't move past a frame until you correct all your mistakes, the space should be highlighted in some way to make it obvious to the user that they typed something else instead.

So for both 1 and 2, I'd love to see these fixed if possible. I don't imagine it will happen all that often, but restricting a user from ever going back to the previous line feels buggy to me. Same deal with number 2, not everyone cares about getting a perfect accuracy result, so forcing them to type a line correctly is something I don't want to enforce.

As for typing an incorrect character on a space, I totally agree! This is a bug I'd like to fix for sure. One approach I hadn't considered that I saw recently in this [typing tui](https://github.com/bloznelis/typioca) is to underline incorrect characters in addition to coloring them red. This works because obviously you can't color a space, but you can if that space is underlined. I'll work on adding this separately though, so you don't have to fuss yourself with it!

Let me know if this all makes sense. I'll see if I can't slot out some time to pair on this as well if you'd like!

TheTrio committed:
Remove redundant condition

I just realized that even when the number of characters is equal to the max length, the next word won't fit on that line. So there's no need to treat this case separately

TheTrio committed:
allow scrolling even when entering incorrect characters

TheTrio committed:
Allow going back to previous lines

TheTrio commented:
> So for both 1 and 2, I'd love to see these fixed if possible. I don't imagine it will happen all that often, but restricting a user from ever going back to the previous line feels buggy to me. Same deal with number 2, not everyone cares about getting a perfect accuracy result, so forcing them to type a line correctly is something I don't want to enforce.

This is how it works now - 

![demo](https://user-images.githubusercontent.com/10794178/170871712-f349d7c9-e63e-47eb-bca1-7257db659563.gif)

I have swapped out the Deque in favor of a Vector since we need to keep track of the lines even after the user is past them(since they can backspace into them again). I've also added a set which keeps tracks of the end points of each line to make it easier to know when the user has backspaced into the previous line. 

> Let me know if this all makes sense. I'll see if I can't slot out some time to pair on this as well if you'd like!

Great! I think I have a working version, but there might be some issues I haven't caught or considered.

TheTrio committed:
Merge branch 'main' into develop

coloradocolby commented:
@TheTrio this is looking great! I don't have time today to give it another review, but it's definitely on my todo list to get this wrapped up and merged this week! Thanks again for all your hard work! 🍻

TheTrio committed:
Merge branch 'main' into develop

TheTrio committed:
Refactored

TheTrio commented:
So replacing the space with a dot might have unintended consequences since spaces are used to determine when to wrap lines. 

![error](https://user-images.githubusercontent.com/10794178/171073761-7ec2120f-5e6c-4b0f-9f19-0a4174a89ff6.gif)

(this is on main btw)

I think underlining is a better option :shrug:

TheTrio committed:
Centered prompt lines

coloradocolby commented:
> So replacing the space with a dot might have unintended consequences since spaces are used to determine when to wrap lines.
> 
> ![error](https://user-images.githubusercontent.com/10794178/171073761-7ec2120f-5e6c-4b0f-9f19-0a4174a89ff6.gif) [ ](https://user-images.githubusercontent.com/10794178/171073761-7ec2120f-5e6c-4b0f-9f19-0a4174a89ff6.gif) ![error](https://user-images.githubusercontent.com/10794178/171073761-7ec2120f-5e6c-4b0f-9f19-0a4174a89ff6.gif) [ ](https://user-images.githubusercontent.com/10794178/171073761-7ec2120f-5e6c-4b0f-9f19-0a4174a89ff6.gif)
> 
> (this is on main btw)
> 
> I think underlining is a better option 🤷

I decided against underlining because I personally felt it made the virtual cursor (which just underlines your current position) have less visual importance and get lost in the midst of any mistakes you make. It seems like we should be able to easily transition your logic to adjust for this. I can take a look myself!

TheTrio commented:
> I decided against underlining because I personally felt it made the virtual cursor (which just underlines your current position) have less visual importance and get lost in the midst of any mistakes you make

Hmm I guess that's true. 

> It seems like we should be able to easily transition your logic to adjust for this. I can take a look myself!

Great!

jrnxf commented:
@TheTrio would you mind leaving some comments on some of the methods you made? I tried looking over this last night and struggled to make heads or tails on some of the methods. I'd definitely love to workshop this together and get this functionality added!

TheTrio committed:
add comments

TheTrio commented:
Hi! I've added some comments. Hopefully makes more sense. Been a while since I touched this code.

I still believe the simplest way to fix this issue would be to not alter the space character - like I suggested earlier, perhaps an underline. There might be a way to get this feature working without that - its just that I can't think of any.

TheTrio commented:
Don't think there's much left to do here, closing. Thanks for all your help!
